docs/rfcs/0003-initial.md

```diff
@@ -0,0 +1,280 @@
+    - Ends the transaction, with any changes rolled back. The transaction object
+      will not be used again.
+  - `.startTransaction[Sync](): driver.Transaction`
+    - Starts a new child transaction within this transaction. The transaction
+      object will not be used again until the child transaction has ended.
+
+### Driver `Meta` Type Information
+
+A driver SHOULD declare an associated `Meta` type using the `driver.Meta<{...}>`
+type function. This is currently used only to specify the `Value` type used by
+the driver for bound and result column values. The `Meta` type SHOULD be
+provided as the (optional) first type argument to every interface that is
+implemented from `driver.sql`.
+
+```ts
+type Meta = driver.Meta<{
+  Value: bigint | number | null
+}>;
+
+export class Driver extends driver.Driver<Meta> { … }
+```
+
+If a driver does not define and use its own `Meta` type, a default `Meta` will
+be used which expects only the the JSON primitive types: `null`, `boolean`,
+`number` and `string`.
+
+## Database Consumer API (`x/database/sql/sql.ts`)
+
+Inspired by [`sql.go`](https://golang.org/src/database/sql/sql.go), this is
+primary interface, which most users will use. `Value` below represents the
+driver-defined `Value` type for bindings and results. The interface is entirely
+async for now, even if the underlying driver supports sync operations.
+
+- `sql.open(path, driver): Promise<sql.Database>`
+  - Creates a database handle/connector with the given driver and path. May
+    validate the arguments (path), but will not open a connection yet.
+- `sql.Database` class
+  - `.connect(): Promise<sql.Connection>`
+    - Opens a new open connection to the database. In the future a connection
+      pool will be maintained instead of always opening new connections.
+- `sql.Connection` class
+  - `.startTransaction(): Promise<sql.Transaction>`
+    - Starts a new transaction in the connection. If a top-level transaction is
+      already in progress on this connection, this will block until it is
+      finished.
+  - `.close(): Promise<void>`
+    - Closes the connection. If there is an active transaction, this will block
+      until it is finished.
+- `sql.Transaction` class
+  - `.prepareStatement(query: string): Promise<sql.PreparedStatement>`
+    - Prepares a SQL query for execution in this transaction.
+  - `.commit(): Promise<void>`
+    - Completes the transaction with changes committed.
+  - `.rollback(): Promise<void>`
+    - Completes the transaction with changes rolled back.
+  - `.startTransaction(): Promise<sql.Transaction>`
+    - Starts a nested transaction within this transaction. If a nested
+      transaction is already in progress, this will block until it is finished.
+      While a nested transaction is in progress, queries should be executed
+      through the inner-most active transaction, not the parent transaction.
+- `sql.PreparedStatement` class
+  - `.query(args?): AsyncGenerator<Iterator<Value>>`
```

Comment:
Not sure if it should be stated, but there should be an option for receiving the data in two formats:

1. as an array of values 

```ts
query("select id, name from table") // [[1,"John"],[2,"Doe"]]
```

2. as an array of objects 

```ts
query("select id, name from table") // [{"id": 1, "name": "John"},{"id": 2, "name": "Doe"}]
```

Reply:
I think we can add a method which returns the column names/ meta for a given query, and then the consumer always has the full information if needed; maybe something like:
```typescript
interface ColumnMeta {
  name: string,
  originName: string?,
  tableName: string?,
};
columnMeta(): Promise<[ColumnMeta]>
```

(SQLite (and the Deno wrapper) exposes a similar api: https://dyedgreen.github.io/deno-sqlite/#/api?id=rowscolumns)

Reply:
@halvardssm I definitely agree that the consumer API should expose something like that, probably implement on top of something like the ColumnMeta @dyedgreen mentions. I'm not sure how I want to handle that in the type system -- Deno Postgres has some interesting ideas to consider -- but maybe we can start with something weakly typed.

Reply:
Added a note about this under "Future Enhancements".